lib/web/app/auth.go

```diff
@@ -39,8 +39,13 @@ type fragmentRequest struct {
 	StateValue         string `json:"state_value"`
 	CookieValue        string `json:"cookie_value"`
 	SubjectCookieValue string `json:"subject_cookie_value"`
+	RequiredApps       string `json:"required_apps"`
 }
 
+// TeleportNextAppRedirectUrlHeader is used to tell the browser which URL to navigate to
+// next in the chain of required app authentication redirects
+const TeleportNextAppRedirectUrlHeader = "X-Teleport-NextAppRedirectUrl"
```

Comment:
Curious, what other alternatives could we have here instead of chaining all the required app auth in redirect? 
I guess this technique will require zero changes on the existing app auth flow. But otherwise what are the downsides to authenticate required apps in a single auth request?

Besides the many redirect dance it needs to perform, this will also create different audit events for all the "required apps" session. We should then also update the app access audit event to write origin of the session so the events can be tied together.

Reply:
> authenticate required apps in a single auth request

The redirect dance goes to the FQDN of the app to set the cookie, so this wouldn't be possible

Reply:
+1 what Ryan said. it was my first thought as well and I tried some initial attempts but the browser must receive the cookie for any of it to work. 

> We should then also update the app access audit event to write origin of the session so the events can be tied together.

i like the idea of adding an origin to the audit events. I'll look into that.

Reply:
> The redirect dance goes to the FQDN of the app to set the cookie, so this wouldn't be possible

I recon this is the current way to set session cookie for app. With some modification to that process, the alternative I was thinking if we can tuck in required apps within parent app as its subdomain.

```
# spec modified to match with the sample provided in PR description
app_service:
  enabled: "yes"
  debug_app: true
  apps:
    - name: "client"
      uri: "http://localhost:3002"
      public_addr: "client.teleport.zarquon.sh"
      required_apps:
        - name: "api22"
          uri: "http://localhost:4000"
          public_addr: "api.client.teleport.zarquon.sh"
```
That way, the required apps won't need to show up as a resource that can be accessed directly and can only be accessed after authenticating with the parent app. 

The auth flow would be something like the following.
1. User launch `client.teleport.zarquon.sh`, performs single MFA tap if per session MFA is enabled (currently it requires MFA tap for every required app).
2. Proxy authenticates the app session request, set's a single auth event that lists authorized subdomains (the required apps) and sets cookie that works for `*.client.teleport.zarquon.sh`
3. On subsequent request, proxy will authenticate subdomains based on cookie set set for `*.client.teleport.zarquon.sh` and the "required apps" spec as a whitelist reference.

Reply:
i suppose this would work strictly for "this app only exists to satisfy client.teleport.com" but if `api.teleport.com` satisfies requests for `client.teleport.com` and `client2.teleport.com` etc etc, this would break down.

Reply:
Umm you mean same app that may be required for two different apps? We can still support the following config no? 
```
apps:
 - name: "api22"
   uri: "http://localhost:4000"
   type: "cors" # specify that app is intended to be used only as required app.
 - name: "client"
   uri: "http://localhost:3002"
   public_addr: "client.teleport.zarquon.sh"
   required_apps:
     - name: "api22"
       public_addr: "api.client.teleport.zarquon.sh"
 - name: "client2"
   uri: "http://localhost:3003"
   public_addr: "client2.teleport.zarquon.sh"
   required_apps:
     - name: "api22"
       public_addr: "api.client2.teleport.zarquon.sh"
```

Reply:
Would that mean we'd have to define the public addrs on api22's to match the public addr for any app we've set it as required?

Reply:
In the example above, yes. Or, since we control the proxy, we might alter the requests to the required apps on the fly so it always gets forwarded to a single public addr on api22. I am still speaking out of my mind and there may be edge cases where this approach could be an issue. But I think this is worth considering given how it simplifies the auth flow.

Reply:
Thanks. I'll experiment a bit with that approach (or something similar).

Reply:
After chatting with the team and review from doyensec, we decided this current implementation is preferred over doing the subdomain way. Could you take another look when you get the chance with that in mind @flyinghermit ? Thank you!